kernel/data-pipeline/core/src/main/java/org/apache/shardingsphere/data/pipeline/core/importer/sink/PipelineDataSourceSink.java

```diff
@@ -283,7 +251,30 @@ private void sequentialFlush(final DataSource dataSource, final List<DataRecord>
             return;
         }
         try (Connection connection = dataSource.getConnection()) {
-            doFlush(connection, buffer);
+            // TODO it's better use transaction, but execute delete maybe not effect when open transaction of PostgreSQL sometimes
+            for (DataRecord each : buffer) {
+                switch (each.getType()) {
+                    case IngestDataChangeType.INSERT:
+                        if (null != rateLimitAlgorithm) {
+                            rateLimitAlgorithm.intercept(JobOperationType.INSERT, 1);
+                        }
+                        executeBatchInsert(connection, Collections.singletonList(each));
+                        break;
+                    case IngestDataChangeType.UPDATE:
+                        if (null != rateLimitAlgorithm) {
+                            rateLimitAlgorithm.intercept(JobOperationType.UPDATE, 1);
+                        }
+                        executeUpdate(connection, each);
+                        break;
+                    case IngestDataChangeType.DELETE:
+                        if (null != rateLimitAlgorithm) {
+                            rateLimitAlgorithm.intercept(JobOperationType.DELETE, 1);
+                        }
+                        executeBatchDelete(connection, Collections.singletonList(each));
+                        break;
+                    default:
+                }
```

Comment:
Could we just invoke `doFlush(Collections.singletonList(each))`? Looks there's duplicated code.